Plugins/Effect/NWScript/nwnx_effect.nss

```diff
@@ -90,16 +91,40 @@ object NWNX_Effect_GetEffectExpiredCreator();
 /// @return Number of internal effects updated.
 int NWNX_Effect_ReplaceEffect(object obj, effect eOld, effect eNew);
 
-/// @}
+/// @brief Gets the true effect count
+/// @param oObject The object to get the count of.
+/// @return the number of effects (item properties and other non-exposed effects included)
+int NWNX_Effect_GetTrueEffectCount(object oObject);
 
-struct NWNX_EffectUnpacked NWNX_Effect_UnpackEffect(effect e)
-{
-    string sFunc = "UnpackEffect";
+/// @brief Gets a specific effect on an object. This can grab effects normally hidden from developers, such as item properties.
+/// @param oObject The object with the effect
+/// @param nElement The point in the array to retrieve (0 to GetTrueEffectCount())
+/// @return A constructed NWNX_EffectUnpacked.
+struct NWNX_EffectUnpacked NWNX_Effect_GetTrueEffect(object oObject, int nElement);
 
-    NWNX_PushArgumentEffect(NWNX_Effect, sFunc, e);
-    NWNX_CallFunction(NWNX_Effect, sFunc);
+/// @brief Replaces an already applied effect with another.
+/// @param oObject The object with the effect to replace
+/// @param nElement The array element to be replaced
+/// @param e The unpacked effect to replace it with.
+/// @note Cannot replace an effect with a different type.
+void NWNX_Effect_ReplaceEffectByElement(object oObject, int nElement, struct  NWNX_EffectUnpacked e);
+
+/// @brief Removes effect by ID
+/// @param oObject The object to remove the effect from
+/// @param sID The id of the effect, can be retrieved by unpacking effects.
+/// @return FALSE/0 on failure TRUE/1 on success.
+int NWNX_Effect_RemoveEffectById(object oObject,  string sID);
+
+/// @brief Disables effect-level immunities.
+/// @param nEnable set to true to turn on.
+void NWNX_Effect_SetEffectImmunityBypass(int nEnable);
```

Comment:
This seems better suited for a Tweak? Is this actually something you expect people to toggle on/off constantly at runtime?

Reply:
Answering this one new given it's more philosophical than coding, get to the coding in a couple days hopefully:

I entertained making this one a tweak when it was initially coded but decided against for 2 reasons:

1) Most (assumption there :P) nwnx users are working with established modules and likely have allowed immunities to levels they find satisfactory. But  may wish to allow only a few effects to bypass immunity. Instead of allowing all effects to bypass immunities and then having to go to each nwscript individually to add a GetHasEffectImmunity call.

2) Some immunities don't have clear events available on the nwscript side, some examples being sneak attack, crippling strike, critical hits, death attack - paralyze. Meaning if you disabled all immunities these would always proc; as long as their individual conditions are met.

2b) Some of these may be workable with nwnx events; but i'm more comfortable with saying "if you want crippling strike to be able to bypass immunity you can use xyz event with this function" instead of "if you want to return to original functionality without disabling the tweak for crippling strike, you can enable the events plugin and use xyz event and skip it. when these conditions are met"

That said i do think there's a flaw in it's current implementation, I've had it running in onEquip before and nwscript onequip (i didn't have a chance to try the after yet) and immunities weren't toggleing back on fast enough; some sneak attacks got through to creatures that should be immune and the like. So I'm thinking about allowing people to specify the object it's working for, or the effect, if not both (and likely it'd be both) so it wouldn't interfere with other players on the server.

Reply:
That would be a lot more useful. Perhaps something like
```C
// override: 0 - not immune, 1 - immune, -1 use efault 
NWNX_Creature_SetImmunityOverride(object creature, int immunity_type, int override);
```